components/ws-proxy/pkg/sshproxy/server.go

```diff
@@ -0,0 +1,170 @@
+// Copyright (c) 2021 Gitpod GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License-AGPL.txt in the project root for license information.
+
+package sshproxy
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"strings"
+	"time"
+
+	supervisor "github.com/gitpod-io/gitpod/supervisor/api"
+	p "github.com/gitpod-io/gitpod/ws-proxy/pkg/proxy"
+	"golang.org/x/crypto/ssh"
+	"golang.org/x/xerrors"
+	"google.golang.org/grpc"
+)
+
+type Session struct {
+	Conn                *ssh.ServerConn
+	WorkspaceId         string
+	PublicKey           ssh.PublicKey
+	WorkspaceIp         string
+	WorkspacePrivateKey ssh.Signer
+}
+
+type Server struct {
+	ConnectionTimeout     time.Duration
+	sshConfig             *ssh.ServerConfig
+	workspaceInfoProvider p.WorkspaceInfoProvider
+}
+
+func (s *Server) HandleConn(c net.Conn) {
+	sshConn, chans, reqs, err := ssh.NewServerConn(c, s.sshConfig)
+	if err != nil {
+		c.Close()
+		return
+	}
+	defer sshConn.Close()
+
+	if sshConn.Permissions == nil || sshConn.Permissions.Extensions == nil || sshConn.Permissions.Extensions["workspaceId"] == "" {
+		return
+	}
+	workspaceId := sshConn.Permissions.Extensions["workspaceId"]
+	wsInfo := s.workspaceInfoProvider.WorkspaceInfo(workspaceId)
+	if wsInfo == nil {
+		return
+	}
+	ctx, _ := context.WithTimeout(context.Background(), time.Second*5)
+	key, err := s.GetWorkspaceSSHKey(ctx, wsInfo.IPAddress)
+	if err != nil {
+		return
+	}
+	session := &Session{
+		Conn:                sshConn,
+		WorkspaceId:         workspaceId,
+		WorkspacePrivateKey: key,
+		WorkspaceIp:         wsInfo.IPAddress + ":23001",
+	}
+
+	go func() {
+		for req := range reqs {
+			switch req.Type {
+			case "keepalive@openssh.com":
+				if req.WantReply {
+					req.Reply(true, []byte{})
+				}
+			default:
+				req.Reply(false, []byte{})
```

Comment:
should we log here unexpected types?

Reply:
can we do something with that log? I.e. are we gonna look at it and extract useful information from that?

Reply:
I was thinking to filter out known types to catch something unexpected, i.e. for cases when a user tries in a way we did not expected. But maybe we can just wait for an issue that some ssh feature is not supported.

Reply:
Let's keep it like that.